spec/lang/intrinsic.md

```diff
@@ -237,3 +237,77 @@ impl<M: Memory> Machine<M> {
     }
 }
 ```
+
+These are the intrinsics for atomic memory accesses:
+
+```rust
+impl<M: Memory> Machine<M> {
+    fn eval_intrinsic(
+        &mut self,
+        Intrinsic::AtomicWrite: Intrinsic,
+        arguments: List<(Value<M>, Type)>,
+        ret_ty: Type,
+    ) -> NdResult<Value<M>> {
+        if arguments.len() != 2 {
+            throw_ub!("invalid number of arguments for `Intrinsic::AtomicWrite`");
+        }
+
+        let Value::Ptr(ptr) = arguments[0].0 else {
+            throw_ub!("invalid first argument to `Intrinsic::AtomicWrite`");
+        };
+
+        let (val, ty) = arguments[1];
+
+        let size = ty.size::<M>();
+
+        if !size.bytes().is_power_of_two() {
+            throw_ub!("invalid second argument to `Intrinsic::AtomicWrite`, size not power of two");
+        }
+
+        if size > M::MAX_ATOMIC_SIZE {
+            throw_ub!("invalid second argument to `Intrinsic::AtomicWrite`, size too big");
+        }
+
+        if !is_unit(ret_ty) {
+            throw_ub!("invalid return type for `Intrinsic::AtomicWrite`")
+        }
+       
+        let pty = PlaceType { ty, align: Align::max_for_offset(size).unwrap() };
```

Comment:
`max_for_offset` doesn't look right. This should just be `Align::from_bytes(size.bytes())` or something like that.